File: backend/discovery.js
const { execFile } = require('node:child_process');
const path = require('node:path');

const PY_SCRIPT = path.join(__dirname, 'discovery.py');

/** Parse the sidecar's stdout; an `{error}` payload becomes a thrown Error. */
function parseDiscoveryOutput(stdout) {
  const data = JSON.parse(String(stdout).trim());
  if (data && data.error) throw new Error(data.error);
  return data;
}

/** Real sidecar invoker: spawn `python3 discovery.py <command> [args]`, parse stdout JSON. */
function makeRunner({ python = 'python3', script = PY_SCRIPT, timeoutMs = 30000 } = {}) {
  return (command, ...args) =>
    new Promise((resolve, reject) => {
      execFile(python, [script, command, ...args], { timeout: timeoutMs }, (error, stdout, stderr) => {
        if (error) return reject(new Error((stderr && stderr.trim()) || error.message));
        try { resolve(parseDiscoveryOutput(stdout)); } catch (e) { reject(e); }
      });
    });
}

async function cached(cache, key, produce) {
  const hit = cache.get(key);
  if (hit) return hit;
  const data = await produce();
  cache.set(key, data);
  return data;
}

const getTrending = (run, cache, region) => cached(cache, `trending:${region}`, () => run('trending', region));
const getRelated = (run, cache, videoId) => cached(cache, `related:${videoId}`, () => run('related', videoId));
const getMoods = (run, cache) => cached(cache, 'moods', () => run('moods'));
const getMood = (run, cache, params) => cached(cache, `mood:${params}`, () => run('mood', params));
const getGenreCharts = (run, cache, region) => cached(cache, `genrecharts:${region}`, () => run('genrecharts', region));
const getPlaylist = (run, cache, playlistId) => cached(cache, `playlist:${playlistId}`, () => run('playlist', playlistId));
const getPodcastSearch = (run, cache, query) => cached(cache, `podcast_search:${query}`, () => run('podcast_search', query));
const getPodcast = (run, cache, showId) => cached(cache, `podcast:${showId}`, () => run('podcast', showId));
const getPodcastFresh = (run, cache, queries) =>
  cached(cache, `podcast_fresh:${queries.join('|')}`, () => run('podcast_fresh', ...queries));

/**
 * Shape guard for discovery payloads. Returns a list of problem strings (empty = OK).
 * Used by the unit tests (against the committed fixtures) and the live `check-discovery`
 * script to detect when ytmusicapi's output drifts from what discovery.py expects — the
 * failure mode that would otherwise silently empty the Discover shelves in production.
 */
function validateDiscoveryShape(kind, data) {
  const problems = [];
  const isStr = (v) => typeof v === 'string' && v.length > 0;
  const checkSongs = (songs, label) => {
    if (!Array.isArray(songs)) return void problems.push(`${label}: songs is not an array`);
    if (songs.length === 0) return void problems.push(`${label}: songs is empty`);
    if (!isStr(songs[0].id)) problems.push(`${label}: songs[0].id missing/empty`);
    if (!isStr(songs[0].title)) problems.push(`${label}: songs[0].title missing/empty`);
  };
  switch (kind) {
    case 'trending':
    case 'related':
    case 'mood':
    case 'playlist':
      checkSongs(data && data.songs, kind);
      break;
    case 'moods': {
      const cats = data && data.categories;
      if (!Array.isArray(cats)) { problems.push('moods: categories is not an array'); break; }
      if (cats.length === 0) { problems.push('moods: categories is empty'); break; }
      if (!isStr(cats[0].key)) problems.push('moods: categories[0].key missing/empty');
      if (!isStr(cats[0].title)) problems.push('moods: categories[0].title missing/empty');
      break;
    }
    case 'genrecharts': {
      const cs = data && data.charts;
      if (!Array.isArray(cs)) { problems.push('genrecharts: charts is not an array'); break; }
      if (cs.length === 0) { problems.push('genrecharts: charts is empty'); break; }
      if (!isStr(cs[0].key)) problems.push('genrecharts: charts[0].key missing/empty');
      if (!isStr(cs[0].title)) problems.push('genrecharts: charts[0].title missing/empty');
      break;
    }
    default:
      problems.push(`unknown kind: ${kind}`);
  }
  return problems;
}

/** Shape guard for podcast payloads (mirrors validateDiscoveryShape). Empty list = OK. */
function validatePodcastShape(kind, data) {
  const problems = [];
  const isStr = (v) => typeof v === 'string' && v.length > 0;
  if (!data || typeof data !== 'object') return [`${kind}: not an object`];
  switch (kind) {
    case 'podcast_search': {
      const shows = data.shows;
      if (!Array.isArray(shows)) { problems.push('podcast_search: shows is not an array'); break; }
      if (shows.length === 0) { problems.push('podcast_search: shows is empty'); break; }
      if (!isStr(shows[0].showId)) problems.push('podcast_search: shows[0].showId missing/empty');
      if (!isStr(shows[0].title)) problems.push('podcast_search: shows[0].title missing/empty');
      break;
    }
    case 'podcast': {
      if (!isStr(data.title)) problems.push('podcast: title missing/empty');
      const eps = data.episodes;
      if (!Array.isArray(eps)) { problems.push('podcast: episodes is not an array'); break; }
      if (eps.length === 0) { problems.push('podcast: episodes is empty'); break; }
      if (!isStr(eps[0].videoId)) problems.push('podcast: episodes[0].videoId missing/empty');
      if (!isStr(eps[0].title)) problems.push('podcast: episodes[0].title missing/empty');
      break;
    }
    case 'podcast_fresh': {
      const shelves = data.shelves;
      if (!Array.isArray(shelves)) { problems.push('podcast_fresh: shelves is not an array'); break; }
      if (shelves.length === 0) { problems.push('podcast_fresh: shelves is empty'); break; }
      const withShows = shelves.find((s) => Array.isArray(s.shows) && s.shows.length > 0);
      if (!withShows) { problems.push('podcast_fresh: no shelf has shows'); break; }
      const show = withShows.shows[0];
      if (!isStr(show.showId)) problems.push('podcast_fresh: shows[0].showId missing/empty');
      if (!isStr(show.title)) problems.push('podcast_fresh: shows[0].title missing/empty');
      const ep = show.episodes && show.episodes[0];
      if (!ep || !isStr(ep.videoId)) problems.push('podcast_fresh: shows[0].episodes[0].videoId missing/empty');
      else if (!isStr(ep.title)) problems.push('podcast_fresh: shows[0].episodes[0].title missing/empty');
      break;
    }
    default:
      problems.push(`unknown kind: ${kind}`);
  }
  return problems;
}

module.exports = { parseDiscoveryOutput, makeRunner, getTrending, getRelated, getMoods, getMood, getGenreCharts, getPlaylist, validateDiscoveryShape, getPodcastSearch, getPodcast, getPodcastFresh, validatePodcastShape };

File: backend/discovery.py
#!/usr/bin/env python3
"""Thin CLI over ytmusicapi (unauthenticated). Prints one JSON object to stdout.
Usage: python3 discovery.py <trending REGION | related VIDEO_ID | moods | mood PARAMS | genrecharts REGION | playlist PLAYLIST_ID | podcast_search QUERY | podcast SHOW_ID | podcast_fresh QUERY...>
On failure: prints {"error": "..."} to stderr and exits non-zero."""
import sys
import json
from ytmusicapi import YTMusic


def _thumb(item):
    thumbs = item.get("thumbnails") or []
    return thumbs[-1]["url"] if thumbs else None


def _to_song(item):
    vid = item.get("videoId")
    artists = item.get("artists") or []
    artist = ", ".join(a.get("name", "") for a in artists if a.get("name")) or "Unknown"
    thumbnail = _thumb(item)
    if not thumbnail and vid:  # e.g. get_watch_playlist tracks omit thumbnails → derive from the videoId
        thumbnail = f"https://i.ytimg.com/vi/{vid}/hqdefault.jpg"
    return {
        "id": vid,
        "title": item.get("title"),
        "artist": artist,
        "thumbnail": thumbnail,
        "duration": item.get("duration"),  # "m:ss" string or None
    }


def _songs_from(items):
    return [_to_song(i) for i in items if i.get("videoId") and i.get("title")]


def _to_show(item):
    # search(filter="podcasts") items are {title, browseId, thumbnails, resultType:"podcast"} — NO author.
    return {
        "showId": item.get("browseId"),
        "title": item.get("title"),
        "thumbnail": _thumb(item),
    }


def _author_name(author):
    # get_podcast's author is a dict {name, id}; be defensive if it's ever a bare string / None.
    if isinstance(author, dict):
        return author.get("name")
    return author or None


def _to_episode(ep):
    return {
        "videoId": ep.get("videoId"),
        "title": ep.get("title"),
        "duration": ep.get("duration"),   # "1 hr 49 min" string — parsed on the client
        "date": ep.get("date"),           # raw, DISPLAY-ONLY (unreliable: can be "35K views")
        "description": ep.get("description"),
        "thumbnail": _thumb(ep),
    }


def cmd_trending(yt, region):
    # Unauthenticated get_charts returns chart *playlists* (no song items; top "songs" need auth).
    # videos[0] is the "Trending NN <country>" playlist — resolve it to real tracks.
    charts = yt.get_charts(country=region)
    videos = charts.get("videos") or []
    playlist_id = videos[0].get("playlistId") if videos else None
    if not playlist_id:
        return {"songs": []}
    pl = yt.get_playlist(playlist_id, limit=50)
    return {"songs": _songs_from(pl.get("tracks") or [])}


def cmd_related(yt, video_id):
    wp = yt.get_watch_playlist(videoId=video_id)
    tracks = wp.get("tracks") or []
    deduped = [t for t in tracks if t.get("videoId") != video_id]  # seed is usually item 0
    return {"songs": _songs_from(deduped)}


def cmd_moods(yt):
    cats = yt.get_mood_categories()
    out = []
    for section, items in (cats or {}).items():
        for it in items or []:
            if it.get("params") and it.get("title"):
                out.append({"key": it["params"], "title": it["title"], "section": section})
    return {"categories": out}


def cmd_mood(yt, params):
    playlists = yt.get_mood_playlists(params) or []
    if not playlists:
        return {"title": "", "songs": []}
    first = playlists[0]
    pl = yt.get_playlist(first.get("playlistId"), limit=50)
    title = first.get("title") or pl.get("title") or ""
    return {"title": title, "songs": _songs_from(pl.get("tracks") or [])}


def cmd_genrecharts(yt, region):
    # genres section is US-only; ranked genre chart playlists, each resolved on tap via cmd_playlist.
    charts = yt.get_charts(country=region)
    genres = charts.get("genres") or []
    out = [{"key": g.get("playlistId"), "title": g.get("title")}
           for g in genres if g.get("playlistId") and g.get("title")]
    return {"charts": out}


def cmd_playlist(yt, playlist_id):
    pl = yt.get_playlist(playlist_id, limit=50)
    return {"title": pl.get("title") or "", "songs": _songs_from(pl.get("tracks") or [])}


def cmd_podcast_search(yt, query):
    results = yt.search(query, filter="podcasts", limit=20) or []
    shows = [_to_show(r) for r in results if r.get("browseId") and r.get("title")]
    return {"shows": shows}


def cmd_podcast(yt, show_id):
    pod = yt.get_podcast(show_id, limit=50)
    episodes = [_to_episode(e) for e in (pod.get("episodes") or []) if e.get("videoId")]
    return {
        "showId": show_id,
        "title": pod.get("title") or "",
        "author": _author_name(pod.get("author")),
        "description": pod.get("description"),
        "thumbnail": _thumb(pod),
        "episodes": episodes,   # newest-first, as ytmusicapi returns them
    }


PODCAST_FRESH_SHOWS_PER_TOPIC = 6


def cmd_podcast_fresh(yt, queries):
    # For each topic query: top K shows from search → each show's LATEST episode (episodes[0], reliable
    # per-show). Per-query and per-show failures degrade to fewer items, never crash the whole command.
    shelves = []
    for q in queries:
        items = []
        try:
            results = yt.search(q, filter="podcasts", limit=20) or []
        except Exception:  # noqa: BLE001
            results = []
        shows = [r for r in results if r.get("browseId") and r.get("title")][:PODCAST_FRESH_SHOWS_PER_TOPIC]
        for s in shows:
            sid = s.get("browseId")
            try:
                pod = yt.get_podcast(sid, limit=5)
                eps = [e for e in (pod.get("episodes") or []) if e.get("videoId")]
                if not eps:
                    continue
                items.append({
                    "showId": sid,
                    "title": s.get("title"),
                    "episodes": [_to_episode(eps[0])],   # latest only (newest-first → [0])
                })
            except Exception:  # noqa: BLE001
                continue
        shelves.append({"query": q, "shows": items})
    return {"shelves": shelves}


def main():
    if len(sys.argv) < 2:
        print(json.dumps({"error": "missing command"}), file=sys.stderr)
        sys.exit(2)
    cmd = sys.argv[1]
    yt = YTMusic()  # unauthenticated
    try:
        if cmd == "trending":
            region = sys.argv[2] if len(sys.argv) > 2 else "US"
            out = cmd_trending(yt, region)
        elif cmd == "related":
            out = cmd_related(yt, sys.argv[2])
        elif cmd == "moods":
            out = cmd_moods(yt)
        elif cmd == "mood":
            out = cmd_mood(yt, sys.argv[2])
        elif cmd == "genrecharts":
            region = sys.argv[2] if len(sys.argv) > 2 else "US"
            out = cmd_genrecharts(yt, region)
        elif cmd == "playlist":
            out = cmd_playlist(yt, sys.argv[2])
        elif cmd == "podcast_search":
            out = cmd_podcast_search(yt, sys.argv[2])
        elif cmd == "podcast":
            out = cmd_podcast(yt, sys.argv[2])
        elif cmd == "podcast_fresh":
            out = cmd_podcast_fresh(yt, sys.argv[2:])
        else:
            print(json.dumps({"error": f"unknown command {cmd}"}), file=sys.stderr)
            sys.exit(2)
    except Exception as e:  # noqa: BLE001 — surface any ytmusicapi failure as JSON on stderr
        print(json.dumps({"error": str(e)}), file=sys.stderr)
        sys.exit(1)
    print(json.dumps(out))


if __name__ == "__main__":
    main()
